python/pyarrow/array.pxi

```diff
@@ -3586,6 +3586,156 @@ class FixedShapeTensorArray(ExtensionArray):
         )
 
 
+cdef class VariableShapeTensorArray(ExtensionArray):
+    """
+    Concrete class for variable shape tensor extension arrays.
+
+    Examples
+    --------
+    Define the extension type for tensor array
+
+    >>> import pyarrow as pa
+    >>> tensor_type = pa.variable_shape_tensor(pa.int32(), 2)
+
+    Create an extension array
+
+    >>> shapes = pa.array([[2, 3], [1, 2]], pa.list_(pa.uint32(), 2))
+    >>> values = pa.array([[1, 2, 3, 4, 5, 6], [7, 8]], pa.list_(pa.int32()))
+    >>> arr = pa.StructArray.from_arrays([shapes, values], names=["shape", "data"])
+    >>> pa.ExtensionArray.from_storage(tensor_type, arr)
+    <pyarrow.lib.VariableShapeTensorArray object at ...>
+    -- is_valid: all not null
+    -- child 0 type: fixed_size_list<item: uint32>[2]
+      [
+        [
+          2,
+          3
+        ],
+        [
+          1,
+          2
+        ]
+      ]
+    -- child 1 type: list<item: int32>
+      [
+        [
+          1,
+          2,
+          3,
+          4,
+          5,
+          6
+        ],
+        [
+          7,
+          8
+        ]
+      ]
+    """
+
+    def to_numpy_ndarray(self):
```

Comment:
Oh, I assumed array of arrays was not techincally possible. Will change.

Reply:
BTW this method also has a different return type than the `FixedShapeTensorArray.to_numpy_ndarray` we added, so that's another reason to use a different name (or just remove it)

For FixedShapeTensorArray the idea was that we wanted to be able to return a single nD array, which is possible for fixed shape, but not for variable shape.

Reply:
Let's remove this `to_numpy_ndarray` method?

Reply:
Removed.